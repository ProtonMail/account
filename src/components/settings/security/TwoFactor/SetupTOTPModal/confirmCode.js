import { Component } from 'preact';
import { connect } from 'unistore/full/preact';

import settingsActions from '../../../../../actions/settings';
import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

import styles from './index.css';

export class ConfirmCode extends Component {
    constructor(props) {
        super(props);
        const {
            settings: { setupTOTP: { request: { TOTPCode } = {} } = {} }
        } = props;
        this.state = {
            code: TOTPCode
        };
    }

    componentWillReceiveProps(newProps) {
        const {
            settings: { setupTOTP: { status, error } = {} }
        } = newProps;

        if (status && status !== this.props.settings.setupTOTP.status) {
            if (status === 'success') {
                this.props.onSubmit();
            }
            if (status === 'failure') {
                this.props.onReset(error ? error.message || error : undefined);
            }
            this.setState({ loading: false });
        }
    }

    onSubmit(e) {
        e.preventDefault();
        this.setState({ loading: true });
        this.props.forbidClosure();
        this.props.enableTOTPAction(this.state.code);
    }

    render() {
        return (
            <ModalWrapper
                onSubmit={(e) => {
                    this.onSubmit(e);
                }}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                <ModalContent>
                    <p className={styles.description}>
                        Test your new 2FA method:
                    </p>
                    <div className="form-row">
                        <label htmlFor="verifyCode">Input your code</label>
                        <div>
                            <input
                                onInput={({ target: { value: code } }) => {
                                    this.setState({ code });
                                }}
                                required={true}
                                value={this.state.code}
                                disabled={this.state.loading}
                                type="code"
                                id="verifyCode"
                                placeholder="Code"
                                minLength="6"
                                maxLength="6"
                                autoFocus={true}
                            />
                        </div>
                    </div>
                </ModalContent>
                <ModalFooter>
                    <button type="reset" value="Reset" disabled={this.state.loading}>
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={this.state.loading}>
                        Finish
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}

export default connect('settings', settingsActions)(ConfirmCode);
