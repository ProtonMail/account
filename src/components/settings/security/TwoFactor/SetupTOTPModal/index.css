.grid {
    width: 50%
}

.row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.label {
    margin: auto 0;
    flex: 1;
}

.value {
    flex: 3
}


.result{
    flex: 2;
}

.description {
    flex: 2;
}

.code {
}
