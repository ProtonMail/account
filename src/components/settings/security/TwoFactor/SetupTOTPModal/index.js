import { h } from 'preact';

import Presentation from './presentation';
import SharedSecret from './sharedSecret';
import { steps as RecoveryCodeSteps, beforeDismiss as beforeDismissRecoveryCodeModal } from '../SaveRecoveryCodeModal';
import { steps as ScopeModal, beforeDismiss as beforeDismissScopeModal } from '../../../../auth/ScopeModal/index';
import store from '../../../../../helpers/store';
import actions from '../../../../../actions/settings';
import ConfirmCode from './confirmCode';

export const steps = [
    {
        title: 'Set Up Two Factor Authentication',
        component({ onNextStep, onPreviousStep, message }) {
            return <Presentation onSubmit={onNextStep} onCancel={onPreviousStep} message={message} />;
        }
    },
    ...ScopeModal('password'),
    {
        title: 'Set Up Two Factor Authentication',
        component({ onNextStep, onPreviousStep }) {
            return <SharedSecret onSubmit={onNextStep} onCancel={onPreviousStep} />;
        }
    },
    {
        title: 'Confirm your new method',
        component({ onNextStep, onPreviousStep, onReset, forbidClosure }) {
            return (
                <ConfirmCode
                    onSubmit={onNextStep}
                    onCancel={onPreviousStep}
                    onReset={onReset}
                    forbidClosure={forbidClosure}
                />
            );
        }
    },
    ...RecoveryCodeSteps
];

export const beforeDismiss = (success = false, reset = false) => {
    beforeDismissScopeModal();
    beforeDismissRecoveryCodeModal();
    if (success || !reset) {
        const { resetStoreAction } = actions(store);
        resetStoreAction(store.getState(), ['setupTOTP']);
    }
};
