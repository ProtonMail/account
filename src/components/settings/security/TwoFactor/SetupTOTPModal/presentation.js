import Link from '../../../../ui/Link';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

/**
 * Shows information about the U2F Key usage.
 * @param {Object} props
 * @param {Function} props.onSubmit - triggers the next step.
 * @param {Function} props.onCancel - triggers the previous step.
 * @return {*}
 */
export default ({ onSubmit, onCancel, message }) => (
    <ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}
    >
        {message ? (
            <ModalContent>
                <div>{message}</div>
            </ModalContent>
        ) : (
            <ModalContent>
                <p>
                    This wizard will enable Two Factor Authentication (2FA) on your ProtonMail account. 2FA will make
                    your ProtonMail account more secure so we recommend enabling it.{' '}
                </p>
                <p>
                    <strong>
                        If you have never used 2FA before, we strongly recommend you reading our 2FA Guide first.
                    </strong>
                </p>
                <div>
                    <Link href="https://protonmail.com/support/knowledge-base/two-factor-authentication/">
                        2FA GUIDE
                    </Link>
                </div>
            </ModalContent>
        )}
        <ModalFooter>
            <button type="reset" value="Reset">
                Cancel
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>
);
