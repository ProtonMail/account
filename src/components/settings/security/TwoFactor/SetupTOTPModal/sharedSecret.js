import { Component } from 'preact';
import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import TextButton from '../../../../ui/TextButton';
import QRCode from 'qrcode.react';

import { connect } from 'unistore/full/preact';
import settingsActions from '../../../../../actions/settings';

import styles from './index.css';

export class SharedSecret extends Component {

    componentDidMount() {
        this.props.createSharedSecretAction();
    }

    constructor(props) {
        super(props);
        this.state = {
            showingQRCode: true
        };
    }

    renderSwitchModeButton() {
        return (<p>
            <TextButton
                onClick={() => this.setState({ showingQRCode: !this.state.showingQRCode })}>
                {this.state.showingQRCode ? 'Enter key manually instead' : 'Scan QR code'}
            </TextButton>
        </p>);
    }

    renderQRCode() {
        const { setupTOTP: { request: { qrURI } = {} } = {} } = this.props.settings;

        return (<ModalContent>
            <p>{this.props.message || 'Scan this QR code with your two factor authentication device to set up your account. '}</p>
            {this.renderSwitchModeButton()}
            {qrURI
                ? <QRCode value={qrURI} renderAs={'svg'} fgColor={'#505061'} size={256}/>
                : <p>Loading...</p>
            }
        </ModalContent>);

    }

    renderRawInformation() {
        const { setupTOTP: { request: { interval, digits, secret } = {} } = {} } = this.props.settings;

        return <ModalContent>
            <p>{this.props.message || 'Manually enter this information into your two factor authentication device to set up your account. '}</p>
            {this.renderSwitchModeButton()}

            {secret ?
                (<div class={styles.grid}>
                    <div class={styles.row}>
                        <label class={styles.label}>KEY</label>
                        <span class={styles.value}><pre>{secret}</pre></span>
                    </div>
                    <div class={styles.row}>
                        <label class={styles.label}>INTERVAL</label>
                        <span class={styles.value}><pre>{interval} seconds</pre></span>
                    </div>
                    <div class={styles.row}>
                        <label class={styles.label}>LENGTH</label>
                        <span class={styles.value}><pre>{digits} digits</pre></span>
                    </div>
                </div>)
                : <p>Loading...</p>
            }
        </ModalContent>;
    }

    render() {
        return (
            <ModalWrapper
                onSubmit={(e) => {
                    e.preventDefault();
                    this.props.onSubmit();
                }}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                {this.state.showingQRCode ? this.renderQRCode() : this.renderRawInformation()}
                <ModalFooter>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit">
                        Next
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}

export default connect('settings', settingsActions)(SharedSecret);
