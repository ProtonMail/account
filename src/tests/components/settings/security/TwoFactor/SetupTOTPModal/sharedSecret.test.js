import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

import { SharedSecret } from '../../../../../../components/settings/security/TwoFactor/SetupTOTPModal/sharedSecret';

describe('SetupTOTP SharedSecret step', () => {
    test('regular display', () => {
        expect(
            render(
                <SharedSecret
                    settings={{
                        setupTOTP: {
                            request: {
                                qrURI: 'otp://it does not really care'
                            }
                        }
                    }}
                />
            )
        ).toMatchSnapshot();

        expect(render(<SharedSecret settings={{}} />)).toMatchSnapshot();
    });

    test('raw display', () => {
        const createSharedSecretAction = jest.fn();

        const tree = deep(
            <SharedSecret
                settings={{
                    setupTOTP: {
                        request: {
                            interval: 60,
                            digits: 25.3,
                            secret: 'THIS_IS_SECRET',
                            qrURI: 'otp://it does not really care'
                        }
                    }
                }}
                createSharedSecretAction={createSharedSecretAction}
            />
        );
        tree.find('a')
            .first()
            .simulate('click');
        expect(tree).toMatchSnapshot();
        expect(createSharedSecretAction).toHaveBeenCalledTimes(1);

        const groot = deep(<SharedSecret settings={{}} createSharedSecretAction={createSharedSecretAction} />);
        groot
            .find('a')
            .first()
            .simulate('click');
        expect(groot).toMatchSnapshot();
        expect(createSharedSecretAction).toHaveBeenCalledTimes(2);
    });

    test('submit and cancel', () => {
        const createSharedSecretAction = jest.fn();

        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const context = deep(
            <SharedSecret
                settings={{}}
                onSubmit={onSubmit}
                onCancel={onCancel}
                createSharedSecretAction={createSharedSecretAction}
            />
        );

        const event = { preventDefault: () => undefined };

        context.find('form').simulate('submit', event);
        expect(onSubmit).toHaveBeenCalled();
        expect(onCancel).not.toHaveBeenCalled();

        context.find('form').simulate('reset', event);
        expect(onSubmit).toHaveBeenCalledTimes(1); // the submit event
        expect(onCancel).toHaveBeenCalled(); // only once

        expect(createSharedSecretAction).toHaveBeenCalledTimes(1);
    });
});
