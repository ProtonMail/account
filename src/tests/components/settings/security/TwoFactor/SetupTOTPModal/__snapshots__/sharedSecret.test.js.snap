// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupTOTP SharedSecret step raw display 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p>
      Manually enter this information into your two factor authentication device to set up your account. 
    </p>
    <p>
      <button
        type="button"
        onClick={[Function onClick]}
        class="button"
      >
        Scan QR code
      </button>
    </p>
    <div class="grid">
      <div class="row">
        <label class="label">KEY</label>
        <span class="value">
          <pre>THIS_IS_SECRET</pre>
        </span>
      </div>
      <div class="row">
        <label class="label">INTERVAL</label>
        <span class="value">
          <pre>60 seconds</pre>
        </span>
      </div>
      <div class="row">
        <label class="label">LENGTH</label>
        <span class="value">
          <pre>25.3 digits</pre>
        </span>
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
    >
      Next
    </button>
  </div>
</form>

`;

exports[`SetupTOTP SharedSecret step raw display 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p>
      Manually enter this information into your two factor authentication device to set up your account. 
    </p>
    <p>
      <button
        type="button"
        onClick={[Function onClick]}
        class="button"
      >
        Scan QR code
      </button>
    </p>
    <p>Loading...</p>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
    >
      Next
    </button>
  </div>
</form>

`;

exports[`SetupTOTP SharedSecret step regular display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p>Scan this QR code with your two factor authentication device to set up
      your account.</p>
    <p>
      <button type="button" class="button">Enter key manually instead</button>
    </p>
    <svg shapeRendering="crispEdges" height="256" width="256" viewBox="0 0 25 25">
      <path fill="#FFFFFF" d="M0,0 h25v25H0z"></path>
      <path fill="#505061" d="M0 0h7v1H0zM9 0h2v1H9zM12 0h1v1H12zM14 0h3v1H14zM18,0 h7v1H18zM0 1h1v1H0zM6 1h1v1H6zM8 1h1v1H8zM11 1h2v1H11zM15 1h2v1H15zM18 1h1v1H18zM24,1 h1v1H24zM0 2h1v1H0zM2 2h3v1H2zM6 2h1v1H6zM8 2h4v1H8zM18 2h1v1H18zM20 2h3v1H20zM24,2 h1v1H24zM0 3h1v1H0zM2 3h3v1H2zM6 3h1v1H6zM9 3h3v1H9zM14 3h3v1H14zM18 3h1v1H18zM20 3h3v1H20zM24,3 h1v1H24zM0 4h1v1H0zM2 4h3v1H2zM6 4h1v1H6zM8 4h1v1H8zM11 4h1v1H11zM13 4h1v1H13zM15 4h2v1H15zM18 4h1v1H18zM20 4h3v1H20zM24,4 h1v1H24zM0 5h1v1H0zM6 5h1v1H6zM8 5h2v1H8zM14 5h3v1H14zM18 5h1v1H18zM24,5 h1v1H24zM0 6h7v1H0zM8 6h1v1H8zM10 6h1v1H10zM12 6h1v1H12zM14 6h1v1H14zM16 6h1v1H16zM18,6 h7v1H18zM8 7h2v1H8zM0 8h2v1H0zM3 8h1v1H3zM6 8h2v1H6zM10 8h2v1H10zM13 8h4v1H13zM18 8h3v1H18zM22 8h2v1H22zM0 9h1v1H0zM2 9h2v1H2zM5 9h1v1H5zM7 9h2v1H7zM10 9h1v1H10zM12 9h2v1H12zM16 9h3v1H16zM21 9h1v1H21zM24,9 h1v1H24zM2 10h1v1H2zM6 10h1v1H6zM9 10h2v1H9zM13 10h1v1H13zM17 10h4v1H17zM23,10 h2v1H23zM0 11h2v1H0zM3 11h1v1H3zM7 11h2v1H7zM13 11h1v1H13zM16 11h1v1H16zM19 11h3v1H19zM0 12h1v1H0zM2 12h1v1H2zM4 12h5v1H4zM12 12h2v1H12zM15 12h4v1H15zM21 12h1v1H21zM23,12 h2v1H23zM3 13h3v1H3zM7 13h1v1H7zM9 13h2v1H9zM12 13h1v1H12zM17 13h2v1H17zM21 13h1v1H21zM24,13 h1v1H24zM0 14h1v1H0zM2 14h1v1H2zM5 14h2v1H5zM10 14h2v1H10zM16 14h1v1H16zM24,14 h1v1H24zM1 15h1v1H1zM4 15h1v1H4zM9 15h1v1H9zM13 15h3v1H13zM18 15h1v1H18zM21 15h1v1H21zM24,15 h1v1H24zM0 16h2v1H0zM6 16h3v1H6zM10 16h1v1H10zM12 16h1v1H12zM16 16h5v1H16zM22 16h2v1H22zM8 17h1v1H8zM11 17h2v1H11zM14 17h3v1H14zM20 17h1v1H20zM22 17h1v1H22zM24,17 h1v1H24zM0 18h7v1H0zM8 18h3v1H8zM12 18h1v1H12zM15 18h2v1H15zM18 18h1v1H18zM20 18h2v1H20zM23,18 h2v1H23zM0 19h1v1H0zM6 19h1v1H6zM9 19h1v1H9zM11 19h1v1H11zM15 19h2v1H15zM20 19h3v1H20zM0 20h1v1H0zM2 20h3v1H2zM6 20h1v1H6zM9 20h6v1H9zM16 20h5v1H16zM0 21h1v1H0zM2 21h3v1H2zM6 21h1v1H6zM8 21h2v1H8zM16 21h1v1H16zM20 21h1v1H20zM22 21h1v1H22zM0 22h1v1H0zM2 22h3v1H2zM6 22h1v1H6zM10 22h1v1H10zM13 22h1v1H13zM20 22h3v1H20zM24,22 h1v1H24zM0 23h1v1H0zM6 23h1v1H6zM8 23h3v1H8zM13 23h2v1H13zM16 23h2v1H16zM20 23h1v1H20zM0 24h7v1H0zM8 24h1v1H8zM11 24h1v1H11zM13 24h1v1H13zM15 24h4v1H15zM23,24 h2v1H23z"></path>
    </svg>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </div>
</form>
`;

exports[`SetupTOTP SharedSecret step regular display 2`] = `
<form class="wrapper">
  <div class="content ">
    <p>Scan this QR code with your two factor authentication device to set up
      your account.</p>
    <p>
      <button type="button" class="button">Enter key manually instead</button>
    </p>
    <p>Loading...</p>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </div>
</form>
`;
